### PR TITLE
Fix/td 11789 If client and server charset are not the same, error will be appear in where string condition

### PR DESCRIPTION
**Pull request by wangmm0220**

### CODE REVIEW

**Review by shenglian-zhou — APPROVED** [review]
test?

File: src/client/src/tscSQLParser.c [hunk 1]
```diff
@@ -5034,6 +5040,21 @@ int32_t handleNeOptr(tSqlExpr** rexpr, tSqlExpr* expr) {
   return TSDB_CODE_SUCCESS;
 }
 
+static void convertWhereStringCharset(tSqlExpr* pRight){
+  if(pRight->value.nType != TSDB_DATA_TYPE_BINARY){
+    return;
+  }
+
+  char newData[TSDB_MAX_NCHAR_LEN] = {0};
```
Comment (shenglian-zhou) [inline src/client/src/tscSQLParser.c]: big allocation on stack. the max length is pRight->value.nLen, malloc it then free?

File: src/client/src/tscSQLParser.c [hunk 2]
```diff
@@ -5034,6 +5040,21 @@ int32_t handleNeOptr(tSqlExpr** rexpr, tSqlExpr* expr) {
...
+  int len = 0;
+  if(!taosMbsToUcs4(pRight->value.pz, pRight->value.nLen, newData, TSDB_MAX_NCHAR_LEN, &len)){
+    tscError("nchar in where condition mbsToUcs4 error");
+  }
+  pRight->value.pz = realloc(pRight->value.pz, len);
```
Comment (glzhao89) [inline src/client/src/tscSQLParser.c]: better check return value of realloc. If allocate failure, the original pointer will be overwritten by NULL thus cause memory leak.

**Review by glzhao89 — APPROVED** [review]

**Comment by wangmm0220** [comment]
> test?

test report:https://jira.taosdata.com:18090/display/DEV/TD-11789

Because it is need to change the charset of configure, it is hard add test case in python test case. I just test it in my machine.

**Pull request merged**